src/vmm/src/boot.rs

```diff
@@ -133,17 +131,146 @@ pub fn build_bootparams(
             add_e820_entry(
                 &mut params,
                 mmio_gap_end.raw_value(),
+                // The unchecked_offset_from is safe because:
+                // * mmio_gap_end > mmio_gap_start, validated at the beginning of the function
+                // * mmio_gap_start > himem_start, validated above
+                // therefore, mmio_gap_end > himem_start
                 // The unchecked + 1 is safe because:
                 // * overflow could only occur if last_addr == u64::MAX and mmio_gap_end == 0
                 // * mmio_gap_end > mmio_gap_start, which is a valid u64 => mmio_gap_end > 0
-                last_addr
-                    .checked_offset_from(mmio_gap_end)
-                    .ok_or(Error::HimemStartPastMmioGapEnd)?
-                    + 1,
+                last_addr.unchecked_offset_from(mmio_gap_end) + 1,
                 E820_RAM,
             )?;
         }
     }
 
     Ok(params)
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use linux_loader::bootparam;
+
+    const HIGH_RAM_START: u64 = 0x0010_0000;
+    // First address past 32 bits is where the MMIO gap ends.
+    const MMIO_GAP_END: u64 = 1 << 32;
+    // Size of the MMIO gap.
+    const MMIO_GAP_SIZE: u64 = 768 << 20;
+    // The start of the MMIO gap (memory area reserved for MMIO devices).
+    const MMIO_GAP_START: u64 = MMIO_GAP_END - MMIO_GAP_SIZE;
```

Comment:
How about we define these as ```pub (crate)``` in vmm/src/lib.rs and then just import them instead of redefining them?

Reply:
They're defined a bit different in `vmm` -
```rust
/// First address past 32 bits.
const FIRST_ADDR_PAST_32BITS: u64 = 1 << 32;
/// Size of the MMIO gap.
const MEM_32BIT_GAP_SIZE: u64 = 768 << 20;
/// The start of the memory area reserved for MMIO devices.
const MMIO_MEM_START: u64 = FIRST_ADDR_PAST_32BITS - MEM_32BIT_GAP_SIZE;
```
and I thought this way was more straightforward. Would it be OK to redefine the ones in `vmm/lib.rs` as `MMIO_GAP_START`, `_SIZE` and `_END` instead of the above, and export them?

Reply:
I would use the values from vmm, and rename them if needed. Defining the constants twice is a bit confusing IMHO.

Reply:
Done